bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java

```diff
@@ -519,9 +518,13 @@ private static boolean isConvertibleToRecordType(Object sourceValue, BRecordType
             }
 
             if ((!returnVal) && (errors.size() >= MAX_CONVERSION_ERROR_COUNT + 1)) {
+                unresolvedValues.remove(typeValuePair);
                 return false;
             }
         }
+        if (!returnVal) {
```

Comment:
We can move this logic to a separate function (`removeTypeValuePair()`) and call it in both places. For positive and negative returns.

Reply:
If we move it to a separate method, the check `!returnVal` will be called twice for the cases where the error message count is passing 20.